File: app/src/main/java/com/kooritea/fcmfix/xposed/PowerkeeperFix.java
package com.kooritea.fcmfix.xposed;

import android.content.Context;

import com.kooritea.fcmfix.util.XposedUtils;

import java.lang.reflect.Field;
import java.util.List;

import com.kooritea.fcmfix.libxposed.XC_MethodHook;
import com.kooritea.fcmfix.libxposed.XposedHelpers;

/**
 * MIUI PowerKeeper 电源管理修复模块
 * 
 * 功能说明：
 * MIUI 的 PowerKeeper（电量管家）会限制 GMS 后台活动，导致 FCM 推送无法正常工作。
 * 本模块通过修改 PowerKeeper 的内部配置，解除对 GMS 的限制，确保 FCM 消息正常接收。
 * 
 * 工作原理：
 * 1. 设置 MilletConfig.isGlobal = true，启用全局模式
 * 2. Hook SimpleSettings.getBoolean()，当查询 "gms_control" 时返回 false（不限制 GMS）
 * 3. Hook MilletPolicy 构造函数，修改三个关键列表：
 *    - mSystemBlackList: 从系统黑名单中移除 GMS
 *    - whiteApps: 从白名单应用中移除 GMS（避免特殊处理）
 *    - mDataWhiteList: 将 GMS 添加到数据网络白名单
 * 
 * 适用场景：
 * - MIUI 系统上 GMS 被电源管理限制
 * - FCM 消息延迟或无法接收
 * - 应用后台被 MIUI 强制杀死
 * 
 * @author fcmfix
 */
public class PowerkeeperFix extends XposedModule {
    public PowerkeeperFix(ClassLoader classLoader) {
        super(classLoader);
        this.startHook();
    }

    /**
     * 开始 Hook 操作
     * 修改 MIUI PowerKeeper 的电源管理策略
     */
    protected void startHook(){
        try {
            // 1. 设置 MilletConfig.isGlobal = true，启用全局管理模式
            Class<?> MilletConfig = XposedHelpers.findClassIfExists("com.miui.powerkeeper.millet.MilletConfig", classLoader);
            XposedHelpers.setStaticBooleanField(MilletConfig, "isGlobal", true);
            printLog("Set com.miui.powerkeeper.millet.MilletConfig.isGlobal to true");

            // 2. Hook SimpleSettings.getBoolean()，绕过 GMS 控制限制
            Class<?> Misc = XposedHelpers.findClassIfExists("com.miui.powerkeeper.provider.SimpleSettings.Misc", classLoader);
            printLog("[fcmfix] start hook com.miui.powerkeeper.provider.SimpleSettings.Misc.getBoolean");
            XposedUtils.findAndHookMethod(Misc, "getBoolean", 3, new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam methodHookParam) throws Throwable {
                    // 当查询 "gms_control" 时，返回 false 表示不限制 GMS
                    if("gms_control".equals((String) methodHookParam.args[1])) {
                        printLog("Success: Success: PowerKeeper GMS Limitation. ", true);
                        methodHookParam.setResult(false);
                    }
                }
            });

            // 3. Hook MilletPolicy 构造函数，修改电源管理策略列表
            Class<?> MilletPolicy = XposedHelpers.findClassIfExists("com.miui.powerkeeper.millet.MilletPolicy", classLoader);

            XC_MethodHook methodHook = new XC_MethodHook() {
                protected void beforeHookedMethod(XC_MethodHook.MethodHookParam methodHookParam) throws Throwable {
                    Field[] declaredFields = null;
                    super.afterHookedMethod(methodHookParam);
                    boolean mSystemBlackList = false;
                    boolean whiteApps = false;
                    boolean mDataWhiteList = false;

                    // 检查 MilletPolicy 类中存在的字段
                    for (Field field : MilletPolicy.getDeclaredFields()) {
                        if (field.getName().equals("mSystemBlackList")) {
                            mSystemBlackList = true;
                        } else if (field.getName().equals("whiteApps")) {
                            whiteApps = true;
                        } else if (field.getName().equals("mDataWhiteList")) {
                            mDataWhiteList = true;
                        }
                    }

                    // 从系统黑名单中移除 GMS，允许其后台运行
                    if (mSystemBlackList) {
                        List blackList = (List) XposedHelpers.getObjectField(methodHookParam.thisObject, "mSystemBlackList");
                        blackList.remove("com.google.android.gms");
                        XposedHelpers.setObjectField(methodHookParam.thisObject, "mSystemBlackList", blackList);
                        printLog("Success: MilletPolicy mSystemBlackList.");
                    } else {
                        printLog("Error: MilletPolicy. Field not found: com.miui.powerkeeper.millet.MilletPolicy.mSystemBlackList");
                    }
                    
                    // 从白名单应用中移除 GMS 和相关服务，避免特殊处理导致的问题
                    if (whiteApps) {
                        List whiteAppList = (List) XposedHelpers.getObjectField(methodHookParam.thisObject, "whiteApps");
                        whiteAppList.remove("com.google.android.gms");
                        whiteAppList.remove("com.google.android.ext.services");
                        XposedHelpers.setObjectField(methodHookParam.thisObject, "whiteApps", whiteAppList);
                        printLog("Success: MilletPolicy whiteApps.");
                    } else {
                        printLog("Error: MilletPolicy. Field not found: com.miui.powerkeeper.millet.MilletPolicy.whiteApps");
                    }
                    
                    // 将 GMS 添加到数据网络白名单，确保后台联网权限
                    if (mDataWhiteList) {
                        List dataWhiteList = (List) XposedHelpers.getObjectField(methodHookParam.thisObject, "mDataWhiteList");
                        dataWhiteList.add("com.google.android.gms");

                        XposedHelpers.setObjectField(methodHookParam.thisObject, "mDataWhiteList", dataWhiteList);
                        printLog("Success: MilletPolicy mDataWhiteList.");
                    }

                }
            };
            printLog("[fcmfix] start hook com.miui.powerkeeper.millet.MilletPolicy constructor");
            XposedHelpers.findAndHookConstructor(MilletPolicy, new Object[] {Context.class, methodHook});

        } catch (XposedHelpers.ClassNotFoundError | NoSuchMethodError  e){
            printLog("No Such Method com.android.server.am.ProcessMemoryCleaner.checkBackgroundAppException");
        }
    }
}

File: app/src/main/kotlin/sumicya/fcmself/mods/PowerkeeperFix.kt
package sumicya.fcmself.mods

import android.content.Context

import io.github.libxposed.api.XposedInterface

import sumicya.fcmself.FcmselfModule
import sumicya.fcmself.core.FcmselfLog
import sumicya.fcmself.hook.Hooks
import sumicya.fcmself.hook.Reflect

/**
 * PowerkeeperFix —— MIUI PowerKeeper 对 GMS 的限制修复。
 *
 * - 把 MilletConfig.isGlobal 置为 true；
 * - 让 SimpleSettings.Misc.getBoolean("gms_control", ...) 恒返回 false；
 * - MilletPolicy 构造时：从 mSystemBlackList 与 whiteApps 移除 GMS（及 ext.services）、
 *   把 GMS 加入 mDataWhiteList。
 *
 * 行为说明：whiteApps 的移除沿用旧版（fcmfix 时代实测保留）逻辑，该列表的语义
 * 未被证实——若实为「允许后台的白名单」，此操作反而收紧。真机验证前不改。
 *
 * 非 MIUI 设备上这些类不存在，各点独立跳过。
 */
internal class PowerkeeperFix(api: XposedInterface, classLoader: ClassLoader) :
    FcmselfModule(api, classLoader, "PowerkeeperFix") {

    init {
        point("MilletConfig.isGlobal") { hookMilletConfig() }
        point("SimpleSettings.Misc#gms_control") { hookMiscGetBoolean() }
        point("MilletPolicy 构造") { hookMilletPolicyConstructor() }
    }

    /** MilletConfig.isGlobal = true */
    private fun hookMilletConfig() {
        val className = "com.miui.powerkeeper.millet.MilletConfig"
        val clazz = Reflect.findClassIfExists(className, classLoader) ?: run {
            skip(className, "类不存在（非 MIUI / HyperOS）")
            return
        }
        Reflect.setStaticObjectField(clazz, "isGlobal", true)
        FcmselfLog.log("Set $className.isGlobal to true")
    }

    /** SimpleSettings.Misc.getBoolean("gms_control", ...) 恒返回 false */
    private fun hookMiscGetBoolean() {
        val className = "com.miui.powerkeeper.provider.SimpleSettings\$Misc"
        val clazz = Reflect.findClassIfExists(className, classLoader) ?: run {
            skip(className, "类不存在（非 MIUI / HyperOS）")
            return
        }
        val method = Reflect.findMethodByParamCount(clazz, "getBoolean", 3)
            ?: throw NoSuchMethodError("$className#getBoolean(3 参)")
        Hooks.hook(api, method) { chain ->
            val result = chain.proceed()
            if (chain.getArg(1) == "gms_control") {
                FcmselfLog.log("Success: PowerKeeper GMS Limitation.", true)
                false
            } else {
                result
            }
        }
    }

    /** MilletPolicy 构造后按字段名调整 GMS 的黑名单 / 白名单归属。 */
    private fun hookMilletPolicyConstructor() {
        val className = "com.miui.powerkeeper.millet.MilletPolicy"
        val clazz = Reflect.findClassIfExists(className, classLoader) ?: run {
            skip(className, "类不存在（非 MIUI / HyperOS）")
            return
        }
        val constructor = Reflect.findConstructorMostMatch(clazz, Context::class.java)
        Hooks.hookAfter(api, constructor) { chain, _ ->
            val policy = chain.thisObject ?: return@hookAfter
            for (field in clazz.declaredFields) {
                when (field.name) {
                    "mSystemBlackList" -> stringList(policy, field.name)?.let { list ->
                        list.remove(GMS)
                        FcmselfLog.log("Success: MilletPolicy mSystemBlackList.")
                    }
                    "whiteApps" -> stringList(policy, field.name)?.let { list ->
                        list.remove(GMS)
                        list.remove(EXT_SERVICES)
                        FcmselfLog.log("Success: MilletPolicy whiteApps.")
                    }
                    "mDataWhiteList" -> stringList(policy, field.name)?.let { list ->
                        if (GMS !in list) list.add(GMS)
                        FcmselfLog.log("Success: MilletPolicy mDataWhiteList.")
                    }
                }
            }
        }
    }

    /** 读一个 List<String> 字段（沿继承链）；类型不符或读取失败返回 null。 */
    private fun stringList(holder: Any, fieldName: String): MutableList<String>? =
        try {
            Reflect.getObjectField(holder, fieldName) as? MutableList<String>
        } catch (_: Throwable) {
            null
        }

    private companion object {
        const val GMS = "com.google.android.gms"
        const val EXT_SERVICES = "com.google.android.ext.services"
    }
}
